File: Grafos/Componentes Conexos/11838 Come and Go.cpp
#include <iostream>
#include <vector>
#include <stack>

#define vvi vector<vector<int> >
#define vi vector<int>
#define si stack<int>
#define endl '\n'

using namespace std;

vi sscInd;
vvi componentes; 
vi enPila;

void tarjan_aux(vvi &grafo, vi &visitados, si &pila, int u, int &t){
    visitados[u] = sscInd[u] = ++t;
    pila.push(u);
    enPila[u] = 1;
    for(int i = 0; i < grafo[u].size(); ++i){
        int v = grafo[u][i];
        if(!visitados[v]){
            tarjan_aux(grafo, visitados, pila, v, t);
            sscInd[u] = min(sscInd[u], sscInd[v]);
        }

        else if(enPila[v] == 1){
            sscInd[u] = min(sscInd[u], sscInd[v]);
        }
    }

    if(sscInd[u] == visitados[u]){
        vi temp;
        bool ver = true;
        while(ver){
            int w = pila.top(); pila.pop();
            if(u == w) ver = false;

            enPila[w] = 0;
            temp.push_back(w);
        }
        componentes.push_back(temp);
    }
}

void tarjan(vvi &grafo){
    int t = 0;
    vi visitados(grafo.size(), 0);
    si pila;
    for(int i = 1; i < visitados.size(); ++i){
        if(!visitados[i]) tarjan_aux(grafo, visitados, pila, i, t);
    }
}

int main(){
    int n, m;
    while(cin >> n >> m, n && m){
        sscInd.clear();
        sscInd.resize(n + 1, 0);
        enPila.clear();
        enPila.resize(n + 1, 0);
        componentes.clear();
        vvi grafo(n + 1);
        for(int i = 0; i < m; ++i){
            int ini, fin, k;
            cin >> ini >> fin >> k;
            grafo[ini].push_back(fin);
            if(k == 2) grafo[fin].push_back(ini);
        }
        
        tarjan(grafo);;
        if(componentes.size() == 1) cout << 1 << endl;

        else cout << 0 << endl;
    }
    return 0;
}

File: Grafos/Componentes Conexos/11838 Come and Go.py
from sys import stdin
from collections import deque

sscInd = []
enPila = []
componentes = []
grafo = []
visitados = []
pila = []
t = 0

def tarjan_aux(u):
    global visitados, sscInd, grafo, componentes, pila, t, enPila
    visitados[u] = t
    sscInd[u] = t
    t += 1
    pila.append(u)
    enPila[u] = 1
    for v in grafo[u]:
        if visitados[v] == 0:
            tarjan_aux(v)
            sscInd[u] = min(sscInd[u], sscInd[v])
        
        elif enPila[v] == 1: sscInd[u] = min(sscInd[u], sscInd[v])

    if sscInd[u] == visitados[u]:
        componentes.append([])
        w = pila.pop()
        while w != u:
            enPila[w] = 0
            componentes[len(componentes) - 1].append(w)
            w = pila.pop()
        
        componentes[len(componentes) - 1].append(u)
        enPila[u] = 0


def tarjan(n):
    global visitados, pila
    visitados = [0 for i in range(n + 1)]
    pila = deque()
    t = 1
    for i in range(1, n + 1):
        if visitados[i] == 0: tarjan_aux(i)


def main():
    global sscInd, enPila, componentes, grafo
    n, m = list(map(int, stdin.readline().split()))
    while n != 0 and m != 0:
        sscInd = [0 for i in range(n + 1)]
        enPila = [0 for i in range(n + 1)]
        componentes = []
        grafo = [[] for i in range(n + 1)]
        for i in range(m):
            ini, fin, k = list(map(int, stdin.readline().split()))
            grafo[ini].append(fin)
            if k == 2: grafo[fin].append(ini)
        
        tarjan(n)
        if len(componentes) == 1: print(1)

        else: print(0)
        
        n, m = list(map(int, stdin.readline().split()))


main()